### DISCUSSION TITLE
BasicDeliverEventArgs.Body - correct usage

### DISCUSSION THREAD

**Discussion by gradx** [rabbitmq/rabbitmq-dotnet-client#1660] [Q&A] [answered]
### Describe the bug

**What version of protobuf and what language are you using?**
Version: 3.28.0
Language: C#

Windows 11
Visual Studio Code
Server: Standalone Docker instance of RabbitMQ 3.13 
Client: RabbitMQ.Client 6.8.1

Method: Any.Parser.ParseFrom().Unpack

### Errors Observed

- Error:Google.Protobuf.InvalidProtocolBufferException: While parsing a protocol message, the input ended unexpectedly in the middle of a field.  This could mean either that the input has been truncated or that an embedded message misreported its own length.
- Error:Google.Protobuf.InvalidProtocolBufferException: Protocol message contained a tag with an invalid wire type.

During testing I noticed only the Client to Server (Protobuf) was erroring and not the Server to Client (Avro) so I switched to using Avro for both and it resolved the issue.  Since I'm not sure where this issue lies I'm submitting the to both the RabbitMQ and Protobuf repositories

### Reproduction steps

1. Serialize message using Protobuf using Any.Pack().ToByteArray()
2. Publish message to excchange 
`var props = pooledObj.Ingress!.CreateBasicProperties();
        props.ReplyTo = command.FireAndForget ? null : pooledObj.EventChannel;
        pooledObj.Ingress.BasicPublish(this.Context.RabbitMQ.Exchange, this.Context.RabbitMQ.RoutingKey, props, bytes);`
3. Deserialize message received from RabbbitMQ
`Any.Parser.ParseFrom(message.Body.ToArray()).Unpack`

### Expected behavior

No Errors

### Additional context

One additional note, of the 10 different brokers I tested only RabbitMQ suffered from this issue.

The source code for this test is available [here](https://github.com/gradx/GenieDotNet/tree/main/GenieDotNet)

**Comment by lukebakken** [comment]
While I appreciate you providing code, I have no idea how to run it to reproduce this behavior, nor do I have time to review such a large amount of code.

If you can demonstrate this issue using a single console application (which should be possible), please do so.

More than likely you are not using this library or RabbitMQ correctly, or something is not working like you expect.

**Comment by gradx** [comment]
Here's a sample log measuring req/sec, 32 connections using .NET crank, 1 hour duration.  If it's not used correctly why are there 6M requests that are successful?

<details><summary>Stats</summary>

```
| load                    |                    |
| ----------------------- | ------------------ |
| Max CPU Usage (%)       | 1                  |
| Max Cores usage (%)     | 23                 |
| Max Working Set (MB)    | 36                 |
| Max Private Memory (MB) | 39                 |
| Build Time (ms)         | 2,197              |
| Start Time (ms)         | 517                |
| Published Size (KB)     | 72,368             |
| Symbols Size (KB)       | 0                  |
| .NET Core SDK Version   | 8.0.401            |
| ASP.NET Core Version    | 8.0.8+954f61dd38b3 |
| .NET Runtime Version    | 8.0.8+08338fcaa5c9 |
| First Request (ms)      | 774                |
| Requests                | 6,757,378          |
| Bad responses           | 82                 |
| Latency 50th (ms)       | 7.78               |
| Latency 75th (ms)       | 9.28               |
| Latency 90th (ms)       | 12.55              |
| Latency 95th (ms)       | 14.51              |
| Latency 99th (ms)       | 17.19              |
| Mean latency (ms)       | 8.50               |
| Max latency (ms)        | 171.89             |
| Requests/sec            | 1,889              |
| Requests/sec (max)      | 3,431              |
| Read throughput (MB/s)  | 0.29               |

[9:33:35 PM] Overall 1871 Period: 1865 Errors: 40
Error:Google.Protobuf.InvalidProtocolBufferException: While parsing a protocol message, the input ended unexpectedly in the middle of a field.  This could mean either that the input has been truncated or that an embedded message misreported its own length.
   at Google.Protobuf.SegmentedBufferHelper.RefillBuffer(ReadOnlySpan`1& buffer, ParserInternalState& state, Boolean mustSucceed)
   at Google.Protobuf.ParsingPrimitives.ReadRawBytesSlow(ReadOnlySpan`1& buffer, ParserInternalState& state, Int32 size)
   at Google.Protobuf.WellKnownTypes.Any.pb::Google.Protobuf.IBufferMessage.InternalMergeFrom(ParseContext& input)
   at Google.Protobuf.CodedInputStream.ReadRawMessage(IMessage message)
   at Genie.IngressConsumer.Services.RabbitMQService.<>c__DisplayClass3_0.<<RabbitMq>b__0>d.MoveNext()
[9:33:48 PM] Overall 1871 Period: 1882 Errors: 41
[9:34:01 PM] Overall 1871 Period: 1880 Errors: 41
[9:34:15 PM] Overall 1871 Period: 1858 Errors: 41
[9:34:28 PM] Overall 1871 Period: 1869 Errors: 41
[9:34:41 PM] Overall 1871 Period: 1882 Errors: 41
Error:Google.Protobuf.InvalidProtocolBufferException: While parsing a protocol message, the input ended unexpectedly in the middle of a field.  This could mean either that the input has been truncated or that an embedded message misreported its own length.
   at Google.Protobuf.SegmentedBufferHelper.RefillBuffer(ReadOnlySpan`1& buffer, ParserInternalState& state, Boolean mustSucceed)
   at Google.Protobuf.ParsingPrimitives.ReadRawBytesSlow(ReadOnlySpan`1& buffer, ParserInternalState& state, Int32 size)
   at Google.Protobuf.WellKnownTypes.Any.pb::Google.Protobuf.IBufferMessage.InternalMergeFrom(ParseContext& input)
   at Google.Protobuf.CodedInputStream.ReadRawMessage(IMessage message)
   at Genie.IngressConsumer.Services.RabbitMQService.<>c__DisplayClass3_0.<<RabbitMq>b__0>d.MoveNext()
[9:34:55 PM] Overall 1871 Period: 1866 Errors: 42
[9:35:08 PM] Overall 1872 Period: 1892 Errors: 42
[9:35:21 PM] Overall 1872 Period: 1874 Errors: 42
[9:35:35 PM] Overall 1871 Period: 1858 Errors: 42
[9:35:48 PM] Overall 1872 Period: 1876 Errors: 42
Error:Google.Protobuf.InvalidProtocolBufferException: While parsing a protocol message, the input ended unexpectedly in the middle of a field.  This could mean either that the input has been truncated or that an embedded message misreported its own length.
   at Google.Protobuf.SegmentedBufferHelper.RefillBuffer(ReadOnlySpan`1& buffer, ParserInternalState& state, Boolean mustSucceed)
   at Google.Protobuf.ParsingPrimitives.ParseRawVarint32SlowPath(ReadOnlySpan`1& buffer, ParserInternalState& state)
   at Google.Protobuf.UnknownFieldSet.MergeFieldFrom(ParseContext& ctx)
   at Google.Protobuf.UnknownFieldSet.MergeFieldFrom(UnknownFieldSet unknownFields, ParseContext& ctx)
   at Google.Protobuf.WellKnownTypes.Any.pb::Google.Protobuf.IBufferMessage.InternalMergeFrom(ParseContext& input)
   at Google.Protobuf.CodedInputStream.ReadRawMessage(IMessage message)
   at Genie.IngressConsumer.Services.RabbitMQService.<>c__DisplayClass3_0.<<RabbitMq>b__0>d.MoveNext()
[9:36:01 PM] Overall 1871 Period: 1866 Errors: 43
[9:<36:15 PM] Overall 1871 Period: 1856 Errors: 43
[9:36:28 PM] Overall 1871 Period: 1878 Errors: 43
Error:Google.Protobuf.InvalidProtocolBufferException: While parsing a protocol message, the input ended unexpectedly in the middle of a field.  This could mean either that the input has been truncated or that an embedded message misreported its own length.
   at Google.Protobuf.SegmentedBufferHelper.RefillBuffer(ReadOnlySpan`1& buffer, ParserInternalState& state, Boolean mustSucceed)
   at Google.Protobuf.ParsingPrimitives.ReadRawBytesSlow(ReadOnlySpan`1& buffer, ParserInternalState& state, Int32 size)
   at Google.Protobuf.WellKnownTypes.Any.pb::Google.Protobuf.IBufferMessage.InternalMergeFrom(ParseContext& input)
   at Google.Protobuf.CodedInputStream.ReadRawMessage(IMessage message)
   at Genie.IngressConsumer.Services.RabbitMQService.<>c__DisplayClass3_0.<<RabbitMq>b__0>d.MoveNext()
Error:Google.Protobuf.InvalidProtocolBufferException: Protocol message contained a tag with an invalid wire type.
   at Google.Protobuf.UnknownFieldSet.MergeFieldFrom(ParseContext& ctx)
   at Google.Protobuf.UnknownFieldSet.MergeFieldFrom(UnknownFieldSet unknownFields, ParseContext& ctx)
   at Google.Protobuf.WellKnownTypes.Any.pb::Google.Protobuf.IBufferMessage.InternalMergeFrom(ParseContext& input)
   at Google.Protobuf.CodedInputStream.ReadRawMessage(IMessage message)
   at Genie.IngressConsumer.Services.RabbitMQService.<>c__DisplayClass3_0.<<RabbitMq>b__0>d.MoveNext()
[9:36:41 PM] Overall 1872 Period: 1885 Errors: 45
[9:36:55 PM] Overall 1872 Period: 1886 Errors: 45
```
</details>

**Reply by lukebakken** [reply]
> "During testing I noticed only the Client to Server (Protobuf) was erroring and not the Server to Client (Avro) so I switched to using Avro for both and it resolved the issue."

Again, you're asking me to spend a significant time debugging something that is probably (even by your own words) not an issue with this library or RabbitMQ.

As I said before, if there is an easy way to reproduce this, even if I have to run something for a long period of time, please provide code and instructions to do so.

**Reply by michaelklishin** [reply]
> Google.Protobuf.InvalidProtocolBufferException

is not a part of this codebase. For RabbitMQ and this client, message bodies are opaque sequences of bytes.

> If it's not used correctly why are there 6M requests that are successful

There could have been be 666M messages, it does not prove anything beyond the fact that it is a behavior that's difficult to reproduce, timing- and data sequence-sensitive issue. That has nothing to do with this client or RabbitMQ in all likelihood.

**Comment by michaelklishin** [comment]
@gradx this client has a ≈ 17 year old history and amongst hundreds of issues discovered, reported and addressed I do not recall any cases of incorrect serialization of AMQP 0-9-1 frames since ≈ 2013.

You must narrow this down to a specific encoded payload and prove that it is this client (or RabbitMQ server) that is somehow responsible for this behavior, which is very likely not the case because message payloads are just byte sequences for this client and RabbitMQ.

It is not up to us to prove that no issue exists. It is up to you to prove its presence. [Only paying users or regular contributors](https://github.com/rabbitmq/rabbitmq-server/blob/main/COMMUNITY_SUPPORT.md) can count on the kind of engagement you seem to be seeking here. I don't recall you being a contributor, and if you have a support subscription, please open a ticket via the support portal.

Even then, we'd ask you to narrow this down to a specific payload

1. Passed on to this Protocol Buffers library (this should be fairly easy to capture), and
2. The AMQP 0-9-1 frame sequence that this client received (this will require taking [a traffic capture](https://www.rabbitmq.com/amqp-wireshark))

"But I can publish millions of messages successfully" proves absolutely nothing except for the fact that this issue is timing- or input sequence-specific.

**Reply by gradx** [reply]
Reproducing the payload is easy since it's Protobuf.

### Definition
https://github.com/gradx/GenieDotNet/blob/main/GenieDotNet/SharedFiles/Protos/shared.proto

### Mock object
https://github.com/gradx/GenieDotNet/blob/main/GenieDotNet/Genie.Common/Web/MockPartyCreator.cs

I'm not here seeking support but rather to report an issue I found while testing various brokers and to bring attention to the matter.  Since you said yourself you haven't seen this issue since 2013 and the community here is a bit defensive I'll let a paying customer find the issue and report it.

**Comment by michaelklishin** [comment]
[This line](https://github.com/gradx/GenieDotNet/blob/741a9e46b34b92e178ef3d130dc028364e67a116/GenieDotNet/Genie.Web.Api/Mediator/Commands/RabbitMQCommand.cs#L33) suggests that publishing is performed with an object pool of some kind.

This very likely means that a shared channel (`IModel` later renamed to `IChannel` or something like that) can be used for concurrent publishing, which [is an absolute no-no with this library](https://www.rabbitmq.com/client-libraries/dotnet-api-guide#concurrency), the docs are pretty explicit about that.

I won't spend any more time guessing because guessing is a very expensive way of troubleshooting distributed systems and binary protocol/codec implementations.

**Reply by gradx** [reply]
> [This line](https://github.com/gradx/GenieDotNet/blob/741a9e46b34b92e178ef3d130dc028364e67a116/GenieDotNet/Genie.Web.Api/Mediator/Commands/RabbitMQCommand.cs#L33) suggests that publishing is performed with an object pool of some kind.
> 
> This very likely means that a shared channel (`IModel` later renamed to `IChannel` or something like that) can be used for concurrent publishing, which [is an absolute no-no with this library](https://www.rabbitmq.com/client-libraries/dotnet-api-guide#concurrency), the docs are pretty explicit about that.
> 
> I won't spend any more time guessing because guessing is a very expensive way of troubleshooting distributed systems and binary protocol/codec implementations.

I'll remove the object pooling and see if it resolves the issue but from testing I believe it was there before it was added.

**Reply by gradx** [reply]
> [This line](https://github.com/gradx/GenieDotNet/blob/741a9e46b34b92e178ef3d130dc028364e67a116/GenieDotNet/Genie.Web.Api/Mediator/Commands/RabbitMQCommand.cs#L33) suggests that publishing is performed with an object pool of some kind.
> 
> This very likely means that a shared channel (`IModel` later renamed to `IChannel` or something like that) can be used for concurrent publishing, which [is an absolute no-no with this library](https://www.rabbitmq.com/client-libraries/dotnet-api-guide#concurrency), the docs are pretty explicit about that.
> 
> I won't spend any more time guessing because guessing is a very expensive way of troubleshooting distributed systems and binary protocol/codec implementations.

[Microsoft.Extensions.ObjectPool](https://learn.microsoft.com/en-us/aspnet/core/performance/objectpool?view=aspnetcore-8.0) is part of the ASP.NET Core infrastructure that supports keeping a group of objects in memory for reuse rather than allowing the objects to be garbage collected. **All the static and instance methods in Microsoft.Extensions.ObjectPool are thread-safe.**

Just an FYI, I removed it to appease you but I was already aware it is threadsafe

**Comment by lukebakken** [comment]
If someone wrote a large C application, that, after running in production for a while, crashed with a segmentation fault, do you think it's reasonable to suspect `gcc`?

> of the 10 different brokers I tested only RabbitMQ suffered from this issue.

Then pick one of these other brokers, and create a project that will demonstrate this issue with RabbitMQ, and that it doesn't happen with one of these other brokers.

**Reply by gradx** [reply]
> If someone wrote a large C application, that, after running in production for a while, crashed with a segmentation fault, do you think it's reasonable to suspect `gcc`?
> 
> > of the 10 different brokers I tested only RabbitMQ suffered from this issue.
> 
> Then pick one of these other brokers, and create a project that will demonstrate this issue with RabbitMQ, and that it doesn't happen with one of these other brokers.

The project already includes every broker and it's easy to toggle between each.  I haven't created the instructions yet because it involves Docker instances but that's next on my list.

So I made the following changes to the Command Handler

        var pooledObj = new RabbitMQPooledObject();
        //var pooledObj = command.GeniePool.Get();

        //pooledObj.Counter++;
        //command.GeniePool.Return(pooledObj);
        pooledObj.Reset();

And performance now suffers tremendously (which was why I added object pooling to begin with) and the issue still remains.

Error:Google.Protobuf.InvalidProtocolBufferException: While parsing a protocol message, the input ended unexpectedly in the middle of a field.  This could mean either that the input has been truncated or that an embedded message misreported its own length.
   at Google.Protobuf.SegmentedBufferHelper.RefillBuffer(ReadOnlySpan`1& buffer, ParserInternalState& state, Boolean mustSucceed)
   at Google.Protobuf.ParsingPrimitives.ReadRawBytesSlow(ReadOnlySpan`1& buffer, ParserInternalState& state, Int32 size)
   at Google.Protobuf.WellKnownTypes.Any.pb::Google.Protobuf.IBufferMessage.InternalMergeFrom(ParseContext& input)
   at Google.Protobuf.CodedInputStream.ReadRawMessage(IMessage message)
   at Genie.IngressConsumer.Services.RabbitMQService.<>c__DisplayClass3_0.<<RabbitMq>b__0>d.MoveNext()

I'll create a project that isolates this issue to RabbitMQ only by tomorrow. 

One other important note.  The Consumer is also multithreaded:

https://github.com/gradx/GenieDotNet/blob/main/GenieDotNet/Genie.IngressConsumer/Services/RabbitMQService.cs

Using TPL Dataflow and a message pump: 

https://github.com/gradx/GenieDotNet/blob/main/GenieDotNet/Genie.Common/Adapters/RabbitMQ/RabbitMQPump.cs

Should I remove this too?

**Reply by lukebakken** [reply]
`IModel` instances are not thread-safe.

If you'd like to assist this project, _and_ possibly resolve this issue, you could update your code to use the latest version 7 RC -

https://www.nuget.org/packages/RabbitMQ.Client/7.0.0-rc.8

`IModel` became `IChannel`, among other API changes, but still remains not-thread-safe. The API is fully async/await, and avoids issues around the thread pool that people reported with version 6. Just remember that `IChannel` is not thread-safe.

If you use version 7.0.0-rc.8, and do _not_ experience this issue, I will buy you a coffee.

**Reply by gradx** [reply]
> `IModel` instances are not thread-safe.
> 
> If you'd like to assist this project, _and_ possibly resolve this issue, you could update your code to use the latest version 7 RC -
> 
> https://www.nuget.org/packages/RabbitMQ.Client/7.0.0-rc.8
> 
> `IModel` became `IChannel`, among other API changes, but still remains not-thread-safe. The API is fully async/await, and avoids issues around the thread pool that people reported with version 6. Just remember that `IChannel` is not thread-safe.
> 
> If you use version 7.0.0-rc.8, and do _not_ experience this issue, I will buy you a coffee.

I upgraded to 7.0.0-rc8 and tested the command handler with and without pooling and the error still persists.  I'll have to create an entirely new RabbitMQService for an  e2e single-threaded test

**Reply by gradx** [reply]
> > `IModel` instances are not thread-safe.
> > If you'd like to assist this project, _and_ possibly resolve this issue, you could update your code to use the latest version 7 RC -
> > https://www.nuget.org/packages/RabbitMQ.Client/7.0.0-rc.8
> > `IModel` became `IChannel`, among other API changes, but still remains not-thread-safe. The API is fully async/await, and avoids issues around the thread pool that people reported with version 6. Just remember that `IChannel` is not thread-safe.
> > If you use version 7.0.0-rc.8, and do _not_ experience this issue, I will buy you a coffee.
> 
> I upgraded to 7.0.0-rc8 and tested the command handler with and without pooling and the error still persists. I'll have to create an entirely new RabbitMQService for an e2e single-threaded test

The single threaded test ran without errors for 15 minutes (20k requests).  What you're proposing basically eliminates the possibility of hosting RabbitMQ in a web service?  I'll let it run overnight for 8 hours to see what happens.

<details><summary>Details</summary>
<p>


| load                    |                    |
| ----------------------- | ------------------ |
| Max CPU Usage (%)       | 0                  |
| Max Cores usage (%)     | 3                  |
| Max Working Set (MB)    | 36                 |
| Max Private Memory (MB) | 31                 |
| Build Time (ms)         | 1,741              |
| Start Time (ms)         | 726                |
| Published Size (KB)     | 72,368             |
| Symbols Size (KB)       | 0                  |
| .NET Core SDK Version   | 8.0.401            |
| ASP.NET Core Version    | 8.0.8+954f61dd38b3 |
| .NET Runtime Version    | 8.0.8+08338fcaa5c9 |
| First Request (ms)      | 660                |
| Requests                | 19,543             |
| Bad responses           | 0                  |
| Latency 50th (ms)       | 198.43             |
| Latency 75th (ms)       | 288.94             |
| Latency 90th (ms)       | 319.07             |
| Latency 95th (ms)       | 386.07             |
| Latency 99th (ms)       | 30,318.64          |
| Mean latency (ms)       | 1,500.45           |
| Max latency (ms)        | 120,141.56         |
| Requests/sec            | 24                 |
| Requests/sec (max)      | 457                |
| Read throughput (MB/s)  | 0.00               |

</p>
</details>

**Comment by gradx** [comment]
[GenieDotNet.zip](https://github.com/user-attachments/files/16793095/GenieDotNet.zip)

I've  attached a simplified project that removed all the unnecessary code.  Once you've installed .NET crank you can run start_crank.bat after  starting the IngressConsumer.

**Reply by lukebakken** [reply]
Thank you, I have added it here - https://github.com/lukebakken/rabbitmq-dotnet-client-1660

I will try to find time to look at this in detail by next Friday.

**Reply by lukebakken** [reply]
@gradx I have invited you as a collaborator in that repository.

https://github.com/lukebakken/rabbitmq-dotnet-client-1660/issues/1

**Comment by lukebakken** [comment]
> What you're proposing basically eliminates the possibility of hosting RabbitMQ in a web service?

I don't know what you mean here.

**Comment by michaelklishin** [comment]
> What you're proposing basically eliminates the possibility of hosting RabbitMQ in a web service?

There's evidence of the contrary that spans 17 years. How you specifically use this client in a Web app may or may not be [optimal](https://www.rabbitmq.com/docs/connections#lifecycle) or fit something else you are trying to do but there's ample evidence that this client is used in .NET-based "Web services."

This client's docs have stated for over decade that [there are known and intentional concurrency limitations](https://www.rabbitmq.com/docs/publishers#concurrency). The entire .NET community was able to deal with this limitation that primarily affects publishing.

**Comment by gradx** [comment, marked as answer]
After working with @lukebakken he helped me discover that BasicDeliverEventArgs [is not thread-safe](https://github.com/lukebakken/rabbitmq-dotnet-client-1660/discussions/2#discussioncomment-10492569). Perhaps BasicDeliverEventArgs .Body should be marked as Obsolete and a descriptor added to indicate how to handle it better similar to how the [RecyclableMemoryStream](https://github.com/microsoft/Microsoft.IO.RecyclableMemoryStream) does so any references display an obvious issue:

```c#
        /// <summary>
        /// Returns a new array with a copy of the buffer's contents. You should almost certainly be using <see cref="GetBuffer"/> combined with the <see cref="Length"/> to
        /// access the bytes in this stream. Calling <c>ToArray</c> will destroy the benefits of pooled buffers, but it is included
        /// for the sake of completeness.
        /// </summary>
        /// <exception cref="ObjectDisposedException">Object has been disposed.</exception>
        /// <exception cref="NotSupportedException">The current <see cref="RecyclableMemoryStreamManager"/>object disallows <c>ToArray</c> calls.</exception>
        /// <exception cref="OutOfMemoryException">The length of the stream is too long for a contiguous array.</exception>
#pragma warning disable CS0809
        [Obsolete("This method has degraded performance vs. GetBuffer and should be avoided.")]
        public override byte[] ToArray()
        {
            this.CheckDisposed();

            string? stack = this.memoryManager.options.GenerateCallStacks ? Environment.StackTrace : null;
            this.memoryManager.ReportStreamToArray(this.id, this.tag, stack, this.length);

            if (this.memoryManager.options.ThrowExceptionOnToArray)
            {
                throw new NotSupportedException("The underlying RecyclableMemoryStreamManager is configured to not allow calls to ToArray.");
            }

            var newBuffer = new byte[this.Length];

            Debug.Assert(this.length <= int.MaxValue);
            this.InternalRead(newBuffer, 0, (int)this.length, 0);

            return newBuffer;
        }
```

**Reply by michaelklishin** [reply]
This is [mentioned in a dedicated .NET client doc guide](https://www.rabbitmq.com/client-libraries/dotnet-api-guide#consuming-memory-safety), although I'd agree that the explanation is quite terse. 

If the idea is to add some code metadata to steer the user to the docs or another method, please submit a PR, especially since you are aware of prior art.

**Reply by gradx** [reply]
> This is [mentioned in a dedicated .NET client doc guide](https://www.rabbitmq.com/client-libraries/dotnet-api-guide#consuming-memory-safety), although I'd agree that the explanation is quite terse.
> 
> If the idea is to add some code metadata to steer the user to the docs or another method, please submit a PR, especially since you are aware of prior art.

If I were to submit a PR, I would add a [MessagePump](https://github.com/gradx/GenieDotNet/blob/main/Genie.Common/Adapters/RabbitMQ/RabbitMQPump.cs)  since during testing I noticed  TPL Dataflow was 2-3x faster than using a thread pool or task. 

- It would abstract the complexity away for the average user and also help mitigate issues like [this](https://github.com/rabbitmq/rabbitmq-dotnet-client/discussions/1635).

- It could solve common threading issues ( also require copying/cloning the entire object) and provide an [easy-to-use ](https://github.com/lukebakken/rabbitmq-dotnet-client-1660/blob/57b3573088fca2153d201fce9d91d92c50c795e4/GenieDotNet/Genie.IngressConsumer/Services/RabbitMQService.cs#L64) solution for consumers.


@lukebakken  mentioned he had some ideas on how to improve the performance so once I hear back from him I'll consider the next steps.

**Reply by michaelklishin** [reply]
@gradx I was only talking about a small doc PR.

**Reply by lukebakken** [reply]
It's not that `.Body` is not thread-safe, it's that the memory that backs `.Body` is a rented array (`ArrayPool`) that is returned immediately after the `Received` event method runs.

This affects all message bodies that are returned by RabbitMQ, not just those exposed via the eventing consumer.

We'll make sure that the documentation in the code blocks and other places are more clear about this.

> If I were to submit a PR, I would add a [MessagePump](https://github.com/gradx/GenieDotNet/blob/main/Genie.Common/Adapters/RabbitMQ/RabbitMQPump.cs) since during testing I noticed TPL Dataflow was 2-3x faster than using a thread pool or task.

Feel free to add an example. In your particular code, you're using a wait event when I think something like a `Channel` would be a better fit, combined with a prefetch value. If I have time to re-visit it, I will.

**Reply by gradx** [reply]
> It's not that `.Body` is not thread-safe, it's that the memory that backs `.Body` is a rented array (`ArrayPool`) that is returned immediately after the `Received` event method runs.

Would it be possible to implement IDisposable and throw a more specific exception when the memory been reclaimed?

**Reply by lukebakken** [reply]
> Would it be possible to implement IDisposable and throw a more specific exception when the memory been reclaimed?

Not without wrapping the underlying memory in some other class, which would increase allocations and decrease performance. That's the sort of thing a higher-level library might do.

I'll make sure the in-line XML docs (and other docs) are clear about this.
